source/common/common/posix/thread_impl.cc

```diff
@@ -31,93 +34,139 @@ int64_t getCurrentThreadId() {
+std::string PosixThread::name() const { return name_; }
 
-  // Thread::Thread
-  void join() override {
-    ASSERT(!joined_);
-    joined_ = true;
-    const int rc = pthread_join(thread_handle_, nullptr);
-    RELEASE_ASSERT(rc == 0, "");
-  }
+// Thread::Thread
+void PosixThread::join() {
+  ASSERT(!joined_);
+  joined_ = true;
+  const int rc = pthread_join(thread_handle_->handle(), nullptr);
+  RELEASE_ASSERT(rc == 0, "");
+}
+
+bool PosixThread::joinable() const { return !joined_; }
+
+ThreadId PosixThread::pthreadId() const {
+#if defined(__linux__)
+  return ThreadId(static_cast<int64_t>(thread_handle_->handle()));
+#elif defined(__APPLE__)
+  uint64_t tid;
+  pthread_threadid_np(thread_handle_->handle(), &tid);
+  return ThreadId(tid);
+#else
+#error "Enable and test pthread id retrieval code for you arch in pthread/thread_impl.cc"
+#endif
+}
 
-private:
 #if SUPPORTS_PTHREAD_NAMING
-  // Attempts to get the name from the operating system, returning true and
-  // updating 'name' if successful. Note that during normal operation this
-  // may fail, if the thread exits prior to the system call.
-  bool getNameFromOS(std::string& name) {
-    // Verify that the name got written into the thread as expected.
-    char buf[PTHREAD_MAX_THREADNAME_LEN_INCLUDING_NULL_BYTE] = {0};
-    const int get_name_rc = pthread_getname_np(thread_handle_, buf, sizeof(buf));
-    name = buf;
-    return get_name_rc == 0;
-  }
+// Attempts to get the name from the operating system, returning true and
+// updating 'name' if successful. Note that during normal operation this
+// may fail, if the thread exits prior to the system call.
+bool PosixThread::getNameFromOS(std::string& name) {
+  // Verify that the name got written into the thread as expected.
+  char buf[PTHREAD_MAX_THREADNAME_LEN_INCLUDING_NULL_BYTE] = {0};
+  const int get_name_rc = pthread_getname_np(thread_handle_->handle(), buf, sizeof(buf));
+  name = buf;
+  return get_name_rc == 0;
+}
 #endif
 
-  std::function<void()> thread_routine_;
-  pthread_t thread_handle_;
-  std::string name_;
-  bool joined_{false};
+class PosixThreadFactoryImpl : public PosixThreadFactory {
+public:
+  ThreadPtr createThread(std::function<void()> thread_routine,
+                         OptionsOptConstRef options) override {
+    return createThread(thread_routine, options, /* crash_on_failure= */ true);
+  };
+
+  PosixThreadPtr createThread(std::function<void()> thread_routine, OptionsOptConstRef options,
+                              bool crash_on_failure) override {
+    auto thread_handle = new ThreadHandle(thread_routine);
+    const int rc = pthread_create(
+        &thread_handle->handle(), nullptr,
+        [](void* arg) -> void* {
+          static_cast<ThreadHandle*>(arg)->routine()();
+          return nullptr;
+        },
+        reinterpret_cast<void*>(thread_handle));
+    if (rc != 0) {
+      delete thread_handle;
+      if (crash_on_failure) {
```

Comment:
Doh, after multiple changes. I realized I had that code removed :( Thanks for pointing it out. I added an early `nullptr`. As for `ENVOY_BUG`, I don't think having a `pthread_create` failure is considered an `ENVOY_BUG`. So, it's probably better for the caller to handle it, whether it can result in an `ENVOY_BUG` or something else.

Reply:
mmm, I think it generally is but realistically we're the only call site so if we don't think it merits an envoy bug for "this is unexpected and we want to understand more" for E-M I'm fine leaving as-is.  @abeyad for thoughts on that

Reply:
I agree that pthread_create failing to create a thread isn't a bug, but from our point of view in Envoy (Envoy and Envoy Mobile), any thread creation that fails (whether it's memory, permissions, something else) is an unexpected occurrence, so I think recording it w/ ENVOY_BUG for potential further analysis makes sense.